### DISCUSSION TITLE
Controlnet is not work with WEBUI Forge

### DISCUSSION THREAD

**Discussion by Dohye111** [lllyasviel/stable-diffusion-webui-forge#563] [Q&A]
I installed WEB UI - FORGE on my computer and attempted to generate an image using Controlnet - Openpose functionality, but Controlnet did not work at all. Judging from the fact that only the picture and skeleton are generated, it seems that Controlnet is working, but the skeleton is not being reflected in the image. 

The Controlnet Model I used is from the models at this link: [,](https://huggingface.co/lllyasviel/sd_control_collection/tree/main) and since these did not work, I also tried models from this link: [,](https://huggingface.co/bdsqlsz/qinglong_controlnet-lllite/tree/main) but still the skeletons of Controlnet are not being reflected in my image.

**Comment by mrsunshine74** [comment]
I have the same problem if you solve it please post the solution thank you!

**Reply by dog0311** [reply]
In my terminal window it doesn't say it's not working it just says it's not enabled. The actual termonolgy it uses is "requires "sd-webui-controlnet" which is disabled".  I have installed it, but I think I screwed up by installing something before installing "controlnet" but I got an error saying I needed to install control net which I did right after that, But I can't figure out how to ENABLE it (ControlNet), I'm not setting some flag somewhere. Maybe it's time to try some of my old debug tricks and see what I can find, worst case, reinstall. Don't know if that helps anyone.

![Screenshot 2025-03-13 202739](https://github.com/user-attachments/assets/54ea10a7-793b-41a9-bb42-754e5abbf060)

**Reply by UnconnectedBedna** [reply]
Enable it then, controlnet does NOT need to be installed, it is integrated in forge, check extensions page and make sure `sd_forge_controlnet` is enabled. `sd_webui_controlnet` is not for forge btw so get rid of that.
You have to download models for the different controlnets though.

**Comment by LucianoDaluz** [comment]
Same here

**Comment by Iter8tion** [comment]
I was also having issues getting Openpose to work. For me, the "preprocessor" button would not work when I was attempting to do "Allow Preview". 
After two days of trying to troubleshoot things and reinstalling twice, I somehow figured out it was my extensions causing an issue.  Meaning that I believed one or more of my extensions were breaking my ability to use OpenPose in ControlNet. 
I did get it to work again. 
What I did was disable all of my extra Extensions (leave the default /built-in Forge ones on).
Then, in groups of 5 extensions....I turned them on, went back to Controlnet, and tried an openpose preview test.  If the Preprocessor button worked and generated a preview in the preview window, then I assumed that those 5 extensions were not creating an issue.
I then went down my entire extension list, adding in 5 extensions, testing, and repeating.  If it suddenly did not function again, then I went back and disabled the last 5 extensions I added, then added each back in one at a time till I determined which one of the 5 was throwing an error.
Once I made it through my list of extensions, it turned out 3 of them affected the preprocessor button (sd-webui-segment-anything, sd-webui-tabs-extension, and stable-diffusion-webui-aesthetic-image-scorer).  Once those 3 were disabled, I was able to use Controlnet with OpenPose again (SD1.5 and SDXL).  
This fixed my particular issue.  It may or may not be an avenue to fix your issue, too.  Good Luck.

**Comment by pflky** [comment]
Controlnet is absolutely wonky in Forge. I get errors for seemingly no reason all the time. The worst part is forge doesn't even allow you to install the default controlnet, which is far more updated. Since stable diffusion is basically nothing without controlnet, this is an issue that makes any speed ups forge gives, pointless....because if you can't use controlnet, you really can't use stable diffusion in general. I've wasted more time trying to get around errors in forge than the amount of extra generation time it would have taken in regular webui.

**Reply by ArnorWing** [reply]
Yes, please allow install of regular CN

**Reply by vixenius** [reply]
Yup, I'm getting errors as well and regular CN just wont enable no matter what i do. Imma stick to a1111 for now since i use cn 99% of the time. p.s. built-in cn wont even match preprocessors and models as it should, its pain in the azz. :(

**Comment by nuclear87** [comment]
Where is the developer >?
i know you already did something great and ppl enjoyed it , dont ruin it as you ignore thier complaining about existing problems !

**Comment by UmbralMoth** [comment]
So to anyone else also looking to use regular ControlNet instead of built-in one. You can just remove regular ControlNet from being always disabled in config.py in modules_forge folder inside the WebUI Forge's main folder.
Just don't forget to disable all the built-in extensions related to built-in ControlNet in extensions!
Edit: Though fair warning, even though it'll enable it, it isn't really intended to be used with Forge, so don't expect it to work, at all.

**Comment by foychermin** [comment]
has anyone come up with a fix for this that doesnt involve diving into py files.........i have no idea whats going on in those files or how they work.

**Comment by mikeharper31** [comment]
Similar problem here.  Can see the list of controlnet pre-processors, but no models at all...
![image](https://github.com/user-attachments/assets/64fe178b-4d0d-4380-8015-1b757f5808f6)

**Reply by changeyi111111** [reply]
I have encountered the same problem as you. I have tried many methods these days, but they do not work. Have you solved it
By the way I'm running on the M2 chip on a mac

**Reply by mikeharper31** [reply]
No, still haven't worked out what the problem is. Tried various things. Posting here was my last hope :)

**Reply by anonymous-person** [reply]
I had that problem too.  The control net models moved from:
  extensions/sd-webui-controlnet/models 
to 
  models/ControlNet
Then they will show up in the model pick list.

But I can only get ControlNet to work if I use a SD1.x model, not Flux.

**Comment by maraan666** [comment]
As mentioned elsewhere on this forum, the rewrite of ControlNet has been delayed. Work will begin around 30th October. Just be patient.

**Reply by Syeph0n** [reply]
Lol finally someone answers with what I was suspecting. Good to know a rewrite is coming! I figured it was something that had been pushed back for a long time.

**Reply by Ultimum007** [reply]
Any news on this issue?

**Reply by CCpt5** [reply]
> Any news on this issue?

Latest date for rewrite of Controlnet (fixes/Flux I assume) is November 20th - see https://github.com/lllyasviel/stable-diffusion-webui-forge/discussions/932#discussioncomment-11121328  
  
I can only imagine the kind of work/studies Forge developer lllyasviel does IRL so I'm sure having specific dates like that set are due to priorities. Many of us are waiitng - it'll be cool when it's back up to date.

**Comment by espritdautomne** [comment]
this fork is just bad and buggy. I just gave it another chance and still, lot of new features, flux support, but man, everything else is broken, no animatediff fix, controlnet is just broken and wont work, 3 out of 4 plugins wont install correctly. This is just a nightmare to use.

**Comment by shirecoding** [comment]
open pose doesnt seem to be working for me

**Comment by UnconnectedBedna** [comment]
Everybody in here should know you most likely have to manually download the models for controlnet, and SD models wont work for SDXL and vice versa.

I have been using models from here.
https://huggingface.co/lllyasviel/ControlNet-v1-1/tree/main
https://huggingface.co/lllyasviel/sd_control_collection/tree/main

Are there better models? Probably, but I have used models from those repos in the past wo problems.
But you have to select the correct model in the dropdown after downloading them. sd model for sd1.5, and SDXL model for SDXL. (use refresh button if they don't appear after placing them in the correct location = `models/ControlNet`)

**Comment by DanielCarmel** [comment]
Any update about this?

**Reply by ArnorWing** [reply]
There is more activity on this issue on the Reforge fork.
Still not working though

**Comment by 0x646e78** [comment]
This may not be relevant for others - I'm using the Stability Matrix AppImage on Linux - but I found the symlinks from the reforge directory were not created, so I had no models available in controlnet. 

I still can't get CN to work `AttributeError: module 'numpy' has no attribute 'bool8'`

But this made the models show up:

```
rm -rf Packages/reforge/models/ControlNet
ln -s /<path to StabilityMatrix>/Data/Models/ControlNet Packages/reforge/models/ControlNet
```

**Reply by UnconnectedBedna** [reply]
~~Why should they? Why should stability matrix create symlinks TO reForge?!?
Besides, what the command you shared does is to create a directory symlink IN ./Packages/reforge/models called ControlNet. (I assume you meant to type `ln` and not `n`)
The syntax is: `ln -s <target> <link_name>`~~

Edit: I misunderstood what stability matrix is. end edit.

Also, this repo is not reForge or stability matrix, it's the original forge.

Edit
The likely reason you have two controlnet directories, IIRC the built-in controlnet in forge uses one and if you clone controlnet (or use the webui interface to download, which means you clone it) you also get the other. I don't remember which is which though. But there is no need to clone controlnet, forge has it built in. All that could be needed is to download models and put in the folder.

**Reply by 0x646e78** [reply]
@UnconnectedBedna  Fair point that this is the original forge, I just happened to find this thread and shared what worked for me. Hence noting my setup may not be relevant to others. I'll think about deleting my comment, maintainers are welcome to anyhow.

I have no huge expectation of Stability Matrix linking things, but they do so to other models such as CheckPoints already, so it could be fair to assume linking ControlNet models is reasonable.

Yes, exactly the 'target' in this case is the place where Stability Matrix downloads it's models to. So, the link example works. A link is created in `reforge/models/` of `ControlNet`, with a target of the StabilityMatrix ControlNet models folder. 

You're probably right about the reason for the second controlnet folder, good point.

**Reply by UnconnectedBedna** [reply]
Since you said "I use the stability matrix appimage" I assumed you meant that is what you are using to run AI. After a search on it, it turns out it is a package manager.
So with that information, yeah, a symlink is required to point to where models are stored.

You don't have to remove anything, leave it in here in case someone, just like you, happens to bump into it when not knowing how symlinks work. (but you should probably edit your initial post with correct information. links are created with `ln`, there is to my knowledge no application called `n`)
Pretty sure this repo is or more or less considered "dead" and no maintainer is going to do anything. Last commit was 11 months ago.
(I still use it though, it works for what I am doing and won't move on until something completely breaks)

**Reply by 0x646e78** [reply]
Good pointers, edited. Yeah I've been finding using things until the completely break is the way. I did get past my 'AttributeError' issue by pinning a bunch of python dep versions. For now, things work.

**Reply by UnconnectedBedna** [reply]
Set up a pyenv and set it to python.3.10 and you should be fine. I do that (python3.10.16) and have zero errors, just a clone of this repo, pyenv, venv and everything works without errors. (IIRC forge even activates the pyenv/venv automatically when starting, it was years since I set it up so I don't remember if you have to activate or not)
https://github.com/pyenv/pyenv
Check the repo of your distro, maybe pyenv is in your repos instead of installing it with curl. On arch that I use there is so I can just `pacman -S pyenv` and then do the stuff in .bashrc (or whatever shell you use) as instructed on the github page.

```
git clone https://github.com/lllyasviel/stable-diffusion-webui-forge.git
cd stable-diffusion-webui-forge
pyenv install 3.10 # will pull 3.10.20 IIRC, I use 3.10.16 but 3.10.20 is probably fine, otherwise use pyenv install 3.10.16
pyenv local 3.10
python -m venv venv

# to activate
. ./venv/bin/activate

# don't forget to deactivate when done if activating
deactivate
```

I use https://github.com/hyperupcall/autoenv to activate/deactivate, but as I said above, not sure that is even needed with forge.
